Implemented `date`

Eike-Flath opened the pull request:
I implemented `date` (issue #6) as described by the [POSIX standard](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/date.html).
I did implement support for changing the system time.
I did not implement any features that are not described in the POSIX standard.

Eike-Flath committed:
Implemented date

Eike-Flath committed:
prevented date from parsing invalid dates

Eike-Flath committed:
use clock_settime (which is specified by POSIX) instead of settimeofday

Sigmanificient commented on src/date/date.c:
```diff
@@ -1,6 +1,119 @@
-}
\ No newline at end of file
+  size_t cap = INITIAL_BUF_SIZE;
+  char *buf = malloc(cap);
```
missing malloc/realloc call checks

Eike-Flath committed:
added malloc/realloc call checks

proh14 commented:
hello thanks for your contribution! please also add `--help` `--version` flags + some of the gnu flags

Eike-Flath committed:
Merge branch 'proh14:main' into date

Eike-Flath committed:
implemented --help and started on --iso-8601

Eike-Flath committed:
Added %N, %:z, --iso-8601 is fully implemented

Eike-Flath committed:
Added RFC options

Eike-Flath commented:
@proh14 I now added `--help` as well as the GNU flags `--iso-8601`, `--rfc-3339`, and `--rfc-email` (as well as their short option equivalents). I also added support for the GNU format specifiers `%N` and `%:z`.
Are there any other GNU features you would like to see implemented?

Sigmanificient commented on src/date/date.c:
```diff
@@ -1,6 +1,338 @@
+#define _POSIX_C_SOURCE 200112L // for setenv
```
You will need to create your own `setenv`. Fear not, it is simpler than it seems.
Take a look at [glibc implementaion](https://sourceware.org/git/?p=glibc.git;a=blob;f=stdlib/setenv.c;h=e2164371ade896e9b3342754724e5ca381beb64e;hb=HEAD) or [muscl one](https://git.musl-libc.org/cgit/musl/tree/src/env/setenv.c)

Sigmanificient commented on src/date/date.c:
```diff
@@ -1,6 +1,338 @@
+  }
+}
+
+static void use_iso_fmt(char *iso) {
+  if (!strcmp(iso, "hours")) {
+    set_operand("+%FT%H%:z");
+  } else if (!strcmp(iso, "minutes")) {
+    set_operand("+%FT%H:%M%:z");
+  } else if (!strcmp(iso, "date")) {
+    set_operand("+%F");
+  } else if (!strcmp(iso, "seconds")) {
+    set_operand("+%FT%T%:z");
+  } else if (!strcmp(iso, "ns")) {
+    set_operand("+%FT%T,%N%:z");
+  } else {
+    fprintf(stderr, "date: invalid argument '%s' for '--iso-8601'\n", iso);
+    fprintf(stderr,
+            "must be one of 'hours', 'minutes', 'date', 'seconds', 'ns'\n");
+    exit(EXIT_FAILURE);
+  }
+}
```
rebase and move to `getopt_long`

Sigmanificient commented on src/date/date.c:
```diff
@@ -1,6 +1,338 @@
+  }
+}
+
+static bool strpre(const char *pre, const char *str) {
+  return !strncmp(pre, str, strlen(pre));
+}
```
Consider
```suggestion
#define strpre(pre, str) (!strncmp((pre), (str), sizeof (pre) - 1))
```
